modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/cache/ClientCachePartitionsRequest.java

```diff
@@ -77,97 +78,68 @@ public ClientCachePartitionsRequest(BinaryRawReader reader) {
             if (cacheDesc == null)
                 continue;
 
-            ClientCachePartitionAwarenessGroup grp = processCache(ctx, groups, cacheGroupIds, affinityVer, cacheDesc);
+            ClientCachePartitionAwarenessGroup grp = processCache(ctx, affinityVer, cacheDesc);
 
-            // Cache already processed.
             if (grp == null)
                 continue;
 
-            groups.add(grp);
-            cacheGroupIds.put(cacheDesc.groupId(), grp);
+            ClientCachePartitionAwarenessGroup grp0 = grps.putIfAbsent(grp, grp);
+
+            if (grp0 != null)
+                grp0.addAll(Collections.singletonList(cacheDesc));
         }
 
-        Map<String, DynamicCacheDescriptor> allCaches = ctx.kernalContext().cache().cacheDescriptors();
+        Map<Integer, List<DynamicCacheDescriptor>> allCaches = ctx.kernalContext().cache().cacheDescriptors().values()
```

Comment:
There is already `ctx.kernalContext().cache().cacheGroupDescriptors()` method, no need to grouping it manually.

Reply:
Discussed privately. Fixed some crutial points.